src/lib/mdns/TxtFields.cpp

```diff
@@ -0,0 +1,281 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "TxtFields.h"
+
+#include <algorithm>
+#include <cstdio>
+#include <inttypes.h>
+#include <limits>
+#include <string.h>
+
+#include <mdns/Resolver.h>
+
+namespace chip {
+namespace Mdns {
+
+namespace Internal {
+
+namespace {
+bool IsKey(const ByteSpan & key, const char * desired)
+{
+    if (key.size() != strlen(desired))
+    {
+        return false;
+    }
+    return memcmp(key.data(), desired, key.size()) == 0;
+}
+
+uint16_t MakeU16FromAsciiDecimal(const ByteSpan & val)
+{
+    uint32_t u32          = 0;
+    const uint16_t errval = 0x0;
+    for (size_t i = 0; i < val.size(); ++i)
+    {
+        char c = static_cast<char>(val.data()[i]);
+        if (c < '0' || c > '9')
+        {
+            return errval;
+        }
+        u32 = u32 * 10;
+        u32 += val.data()[i] - static_cast<uint8_t>('0');
+        if (u32 > std::numeric_limits<uint16_t>::max())
+        {
+            return errval;
+        }
+    }
+    return static_cast<uint16_t>(u32);
```

Comment:
Suggest replacing all the string logic with a call to `strtoul` and simply keeping the limit check. 

This could be done as a TODO, or such. Otherwise, we need a unit test on this conversion routine.

Reply:
Incoming data is not a string.

Reply:
It is a string, just not a null-terminated string.